# TensorFlow: Introduction
*Rachel Buttry*

Updated: *12 April 2018*

### What is TensorFlow?
<img src='./images/TensorFlowLogo.svg' width='25%'>
"TensorFlow is an open-source software library for dataflow programming across a range of tasks. It is a symbolic math library, and is also used for machine learning applications such as neural networks." -[Wikipedia](https://en.wikipedia.org/wiki/TensorFlow)

It allows us to design machine learning models from low level building blocks. This is particularly useful when we want to deep learning.

### Installing TensorFlow
There are many different ways to install tensorflow, all of which can be found [here](https://www.tensorflow.org/install/). Below are the following installations I've gotten to work.

#### Ubuntu:
I installed it using anaconda pretty easily. 
([Instructions listed here](https://www.tensorflow.org/install/install_linux#InstallingAnaconda)):

```
$ conda create -n tensorflow pip python=2.7 # or python=3.3, etc.
$ source activate tensorflow
 (tensorflow)$  # Your prompt should change 
 (tensorflow)$ pip install --ignore-installed --upgrade [tfBinaryURL]

```
for ```[tfBinaryURL]```, copy and past one of the following links:
- [python 2.7](https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-1.7.0-cp27-none-linux_x86_64.whl
)
- [python 3.4](https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-1.7.0-cp34-cp34m-linux_x86_64.whl
)

#### OSX:
On Mac (including tcsh), I was able to install it using virtualenv. ([Instructions listed here](https://www.tensorflow.org/install/install_mac)):
```
$ mkdir tensorflow
$ cd tensorflow
```

Use virtualenv:
```
$ virtualenv --system-site-packages targetDirectory # for Python 2.7
$ virtualenv --system-site-packages -p python3 targetDirectory # for Python 3.n
$ source ./bin/activate      # If using bash, sh, ksh, or zsh
$ source ./bin/activate.csh  # If using csh or tcsh 
```

Install tensorflow in virtualenv:
```
(targetDirectory)$ pip install --upgrade tensorflow      # for Python 2.7
(targetDirectory)$ pip3 install --upgrade tensorflow     # for Python 3.n
```

or if ```pip install --upgrad tensorflow``` doesn't work:
```
 $ pip install --upgrade [tfBinaryURL]   # Python 2.7
 $ pip3 install --upgrade [tfBinaryURL]  # Python 3.n 
```
where ```[tfBinaryURL]``` is the same as in the ubuntu installation

### Validating your  installation

Try running the following cell either here or in a python shell:
```
>>> import tensorflow as tf
```
If you receive the following error:
```
Illegal instruction (core dumped) after running import tensorflow
```
then just downgrade to an earlier tensorflow. To do this, run:
```
(tensorflow)$ pip uninstall tensorflow
(tensorflow)$ pip install tensorflow==1.5
```
You should be able to import tensorflow without any problem now.

### Activating/Using Tensorflow Environment
We can activate the tensorflow environments the same way we did before.

In Ubuntu:
```
$ source activate tensorflow
(tensorflow) $
```
In  Mac (bash):
```
$ cd tensorflow
$ source ./bin/activate
(tensorflow) $
```

In  Mac (tcsh):
```
$ cd tensorflow
$ source ./bin/activate.csh
(tensorflow) $
```

The new prompt means you are now using the tensorflow environemnt. The python shell and any py script you run now will use the tensorflow environment. However any packages you want to use must be installed again while in the tensorflow environment.
```
(tensorflow) $ pip install jupyter
(tensorflow) $ pip install scikit-learn
(tensorflow) $ pip install pandas
```
 So, any jupyter notebook you want to use tensorflow in will have to be opened while in the tensorflow environment.
 
 You can deactivate the tensorflow environment with the command in both bash and tcsh:
```
 $ source deactivate
```

### Quick sidenote:
From here on out, I've assumed you've successfully installed tensorflow and have (re)opened this notebook within the tensorflow environment. We can now do some basic examples.

I reccommend watching [this talk](https://www.youtube.com/watch?v=oxf3o8IbCk4) where I've all of the following examples were inspired by or taken from.

### Example 1: Nodes

In [1]:
# Start by importing tensorflow
import tensorflow as tf

In Tensorflow, nodes serve as operations performed on data and hold parameters we may want to vary. For instance, mathematical operations (addition, subtraction, multiplication, division) look like:
```
tf.add(x,y)
tf.subtract(x,y)
tf.multiply(x,y)
tf.divide(x,y)
```
But you can also just use +,-,\*,/ for shorhand. The long tf form has more arguments that may be helpful depending on what you're doing. These nodes working together fall onto a "graph", more on that in example 2.

In [2]:
# Create two constant nodes 
const1 = tf.constant(3.5)
const2 = tf.constant(10.0)

# Create three nodes to
# find the difference, sqare it, then multiply it by the first constant
diff = tf.subtract(const2, const1)
squared = tf.square(diff)
add = tf.add(const1, squared)

Unlike regular python, we can't just print to see the values of nodes, we have to create a session!

In [3]:
# create a session to use our nodes
sess  = tf.Session()
#or create a remote session
# tf.Session("grpc://example.org:2222")

In [4]:
print sess.run(add)

45.75


Now let's try passing data thru a graph

In [5]:
# Create two placeholders (for two floats)
x  = tf.placeholder(tf.float32)
y  = tf.placeholder(tf.float32)

# Create two nodes that do the following:
# add the values then multiply the value by 3
adder_node = x + y
add_and_triple = adder_node*3

We can pass values through using our session in the format below:

In [6]:
# pass thru sample values
print sess.run(add_and_triple, {x:3, y:4.5})

22.5


(3 + 4.5) \* 3 = 22.5

so it looks like it worked!

But why should we go thru all this trouble when we could just write the simple version of these operations without tensorflow? One benefit is that, once we create our graph, we can pass data in of any shape, so long as the shape of the two data variables are the same!

In [7]:
xvalues = [[666, 10, 3.5], [2, 8, 7]]
yvalues = [[9, 2.1, 4], [5, 2, 6]]
print sess.run(add_and_triple, {x:xvalues, y:yvalues})
sess.close()

[[2025.         36.300003   22.5     ]
 [  21.         30.         39.      ]]


### Example 2: Tensorboard
We can visualize what nodes our data is passing through with tensorboard.

In [8]:
tf.reset_default_graph()#reset graph (or else it will show the example from before)

# create some new nodes
a = tf.constant(5.0, name="input_a")
b = tf.constant(3.0, name="input_b")
c = tf.multiply(a,b, name="multiply_c")
d = tf.add(a,b, name="add_d")
e = tf.add(c,d, name="add_e")

# Run it!
sess = tf.Session()
output = sess.run(e)

# Use the writer to create the graph folder
writer = tf.summary.FileWriter("./my_graph", sess.graph)
writer.close()
sess.close()

We can access the graph we just created by using tensorboard:
```
$ tensorboard --logdir=./my_graph
TensorBoard 1.5.1 at http://acerc720:6006 (Press CTRL+C to quit)
```
Then go to the address in your browser (yours will more than likely not be at acerc720:6006)

This specific graph looks like:
![my graph](./images/my_graph.png)

As computations become more complicated, visualizing your operations on the graph can be a useful tool!

### Example 3: Linear Regression
We can perform a linear fit using a tensorflow. We'll create a linear model: $y = Wx + b$
and then use a gradient descent optimizer to find the ideal parameter values for $W$ and $b$.

In [9]:
#make a linear model
W = tf.Variable([.3], tf.float32)#weight
b = tf.Variable([-.3], tf.float32)#bias
x = tf.placeholder(tf.float32)#data
linear_model = W*x+b

In [10]:
#initialize variables with the values given in first arg
init = tf.global_variables_initializer()
sess = tf.Session()
sess.run(init)

# run our model
print sess.run(linear_model, {x: [1,2,3,4]})

# we just get back W*x+b as we'd expect

[0.         0.3        0.6        0.90000004]


In [11]:
y = tf.placeholder(tf.float32)
squared_deltas = tf.square(linear_model - y) # square difference from y
loss = tf.reduce_sum(squared_deltas) #error function is sum of all the sqared delta values

print "loss: ", sess.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]})

loss:  23.66


We're not done yet! But since this is a really easy system to solve, let's just take a look at what we would get if we used the correct parameters.

In [12]:
#It's a pretty easy system to solve so, let's look at what the correct answer should be
fixW = tf.assign(W, [-1.])
fixb = tf.assign(b, [1.])
sess.run([fixW, fixb])

print "loss: ", sess.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]})

loss:  0.0


Now let's bring in the optimizer!

#### Gradient Descent
Gradient descent is an optimization algorithm that works by moving towards a minimum of the loss function. It calculates the gradient and takes a step in the direction opposite of it.

<img src="./images/Gradient_descent.svg" width="250">
Image from: [Wikipedia](https://en.wikipedia.org/wiki/Gradient_descent)

For our example, we can imagine a 3D plot with $W$,$b$ as the x,y values and the $loss(W,b)$  function as the z axis. If we minimize the loss, we're minimizing the sum of the distances of our fitted line from the data points.

(This is a situation that really doesn't demand the use of machine learning, but it's a good basic optimization example.)

In [13]:
#Now, let's solve it the machine learning way
optimizer = tf.train.GradientDescentOptimizer(0.01) #we'll use gradient descent 
train = optimizer.minimize(loss)

sess.run(init) #reset to intial, incorrect values or variables (W,b)

#train 1000 times
for i in range(1000):
    #each "train" iteration is a step in the negative gradient direction
    sess.run(train, {x:[1,2,3,4], y:[0,-1,-2,-3]})

In [15]:
print sess.run([W,b])
print "loss: ", sess.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]})
sess.close()

 [array([-0.9999969], dtype=float32), array([0.9999908], dtype=float32)]
loss:  5.6999738e-11


It's not exactly the correct answer, but it's pretty damn close. And that's fine, this is the closest we could get with our gradient step size.

### Summary
* Tensorflow is software library we can call using python.
* Nodes are operations, data input, weight parameters, and more!
* The graph is where the nodes live and you can look at the nodes in relation to eachother using tensorgraph.
* Optimize parameters using an optimizer! It may not give you the exact answer, but good enough!

You may have noticed the ```sess.close()``` at the end of each example. This is because each session can only have one graph. A common workaround is to instead use:

```
with sess as tf.Session:
    node = tf.constant(5.0)
    sess.run(node)
```
